clarity/contracts/pool-token/ytp-yield-usda.clar

```diff
@@ -0,0 +1,152 @@
+(impl-trait .trait-ownable.ownable-trait)
+(impl-trait .trait-semi-fungible-token.semi-fungible-token-trait)
+
+(define-constant ERR-NOT-AUTHORIZED (err u1000))
+(define-constant ERR-TOO-MANY-POOLS (err u2004))
+(define-constant ERR-INVALID-BALANCE (err u2008))
+
+(define-fungible-token ytp-yield-usda)
+(define-map token-balances {token-id: uint, owner: principal} uint)
+(define-map token-supplies uint uint)
+(define-map token-owned principal (list 2000 uint))
+
+(define-data-var contract-owner principal tx-sender)
+(define-map approved-contracts principal bool)
+
+(define-read-only (get-owner)
+  (ok (var-get contract-owner))
+)
+
+(define-public (set-owner (owner principal))
+  (begin
+    (asserts! (is-eq contract-caller (var-get contract-owner)) ERR-NOT-AUTHORIZED)
+    (ok (var-set contract-owner owner))
+  )
+)
+
+(define-private (check-is-approved (sender principal))
+  (ok (asserts! (or (default-to false (map-get? approved-contracts sender)) (is-eq sender (var-get contract-owner))) ERR-NOT-AUTHORIZED))
+)
+
+(define-read-only (get-token-owned (owner principal))
+    (default-to (list) (map-get? token-owned owner))
+)
+
+(define-private (set-balance (token-id uint) (balance uint) (owner principal))
+    (begin
+	    (map-set token-balances {token-id: token-id, owner: owner} balance)
+        (map-set token-owned owner (unwrap! (as-max-len? (append (get-token-owned owner) token-id) u2000) ERR-TOO-MANY-POOLS))
```

Comment:
What is the `token-owned` list for? It will make `set-balance` operations more expensive. Also it always appends so you will hit the limit after 2000 iterations, or am I missing something?

Reply:
`token-owned` returns the list of tokens owned by the relevant `owner`. We need a function that returns such a list, so we can iterate on the UI. If you have any other suggestions, more than welcome!

Reply:
I suggest to index these chain events on a backend and query that from the frontend. If the only purpose for the list is display reasons, then we are adding unnecessary runtime and storage cost. As the list grows, calls will become more and more expensive.